### ISSUE TITLE
[BUG] YOLOv8 Live Detection cap.read() Fails on Raspberry Pi 4 with 64-bit OS, Despite libcamera-still Working

### ISSUE THREAD

**Issue by Shamsu992** [ultralytics/ultralytics#21614]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

1. Environment Details

Hardware: Raspberry Pi 4 Model B

Camera: Official Raspberry Pi Camera Module v1.3

Operating System: Raspberry Pi OS (64-bit), based on Debian Bookworm (freshly installed with Raspberry Pi Imager on a new Kingston A1 microSD card).

Software:

Python 3.11 (default with OS)

ultralytics: Installed via pip

torch: Installed via pip from piwheels

opencv-python: Installed via pip

2. Problem Description

I am trying to run a simple, real-time object detection script using a custom YOLOv8s segmentation model for blueberry detection on a Raspberry Pi 4. The script successfully loads the model but fails to capture frames from the camera.

The Python script uses OpenCV (cv2.VideoCapture(0)). The cap.isOpened() check returns True, but the cap.read() command consistently returns False for the success flag, causing the script to exit with my custom error "Error: Failed to grab frame." The script is being run from the full graphical desktop environment with a monitor, keyboard, and mouse connected.

The most confusing part is that the native system-level camera tool, libcamera-still, works perfectly from the command line and successfully captures an image. This suggests the hardware, cable, and base OS drivers are all working correctly.

3. Code to Reproduce

This is the Python script (detect_live.py) being used:

import cv2
from ultralytics import YOLO

# Load the custom trained YOLOv8 model
model = YOLO('best.pt')

# Initialize the camera
cap = cv2.VideoCapture(0)

if not cap.isOpened():
    print("Error: Could not open camera.")
    exit()

# Loop through video frames
while True:
    # Read a frame from the camera
    success, frame = cap.read()

    # If reading the frame fails, print an error and stop.
    if not success:
        print("Error: Failed to grab frame.")
        break

    # Run YOLOv8 inference on the frame
    results = model(frame, conf=0.5)

    # Visualize the results on the frame
    annotated_frame = results[0].plot()

    # Display the annotated frame
    cv2.imshow("Blueberry Detection", annotated_frame)

    # Break the loop if 'q' is pressed
    if cv2.waitKey(1) & 0xFF == ord("q"):
        break

# Release the camera and close the display window
cap.release()
cv2.destroyAllWindows()

4. Troubleshooting Steps Already Taken

Replaced the original microSD card with a brand new, high-quality Kingston A1 card.

Performed a fresh install of the latest 64-bit Raspberry Pi OS.

Ran sudo apt update and sudo apt full-upgrade -y.

Confirmed the camera ribbon cable is securely seated at both ends and in the correct orientation.

Confirmed the camera is detected on boot (rainbow splash screen is visible).

Confirmed libcamera-still -o test.jpg works successfully from the terminal.

Installed all Python packages using pip3 with the --break-system-packages flag.

5. The Core Question

What could cause OpenCV's cv2.VideoCapture(0).read() to fail on a fresh system, when the system's own libcamera-still command works perfectly with the same hardware? Is there a known incompatibility or a missing dependency for the OpenCV/libcamera compatibility layer on the latest Raspberry Pi OS that pip does not install automatically?

### Environment

This is the Python script (detect_live.py) being used:

import cv2
from ultralytics import YOLO

# Load the custom trained YOLOv8 model
model = YOLO('best.pt')

# Initialize the camera
cap = cv2.VideoCapture(0)

if not cap.isOpened():
    print("Error: Could not open camera.")
    exit()

# Loop through video frames
while True:
    # Read a frame from the camera
    success, frame = cap.read()

    # If reading the frame fails, print an error and stop.
    if not success:
        print("Error: Failed to grab frame.")
        break

    # Run YOLOv8 inference on the frame
    results = model(frame, conf=0.5)

    # Visualize the results on the frame
    annotated_frame = results[0].plot()

    # Display the annotated frame
    cv2.imshow("Blueberry Detection", annotated_frame)

    # Break the loop if 'q' is pressed
    if cv2.waitKey(1) & 0xFF == ord("q"):
        break

# Release the camera and close the display window
cap.release()
cv2.destroyAllWindows()

### Minimal Reproducible Example

![Image](https://github.com/user-attachments/assets/94656845-d42b-486b-8266-cbc709cc84fd)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Shamsu992, thank you for reporting this and for your detailed troubleshooting steps! 🚀 This is an automated response to help get you started, and an Ultralytics engineer will review your issue and assist you further soon.

We recommend checking out the [Docs](https://docs.ultralytics.com/) for additional guidance, especially the [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, as well as the [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) guidelines.

As this is a 🐛 Bug Report, please ensure that your example is as minimal and reproducible as possible. If you discover any further details or can simplify your code or steps, please update your report.

Join our Ultralytics community for more support and discussion:
- [Discord](https://discord.com/invite/ultralytics) for real-time chat 🎧
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge

## Upgrade

Please ensure you’re running the latest `ultralytics` package (and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml)) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to rule out issues already fixed in recent releases:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in the following verified environments (with all dependencies, including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** (free GPU): <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your report and patience! An Ultralytics team member will follow up with you soon.

**Comment by Y-T-G** [comment]
You can try this:
https://stackoverflow.com/questions/76819494/my-raspberry-pi-camera-does-not-return-frames-to-opencv/78143323#78143323

Or use a different library to capture:
https://stackoverflow.com/questions/76819494/my-raspberry-pi-camera-does-not-return-frames-to-opencv/76849630#76849630

**Comment by Shamsu992** [comment]
I see, so if I follow that link it will automatically sort out my issue ? Because I have my pt model ready, just want the video to display from the picamera inorder to check how robust my model is and how the confidence threshold works.

I hope this link helps me Muhammad and I will try it. If there is anything you can add to help me I am willing to reply immediately

**Comment by glenn-jocher** [comment]
@Shamsu992 The StackOverflow links should help, but there's a more direct solution for Raspberry Pi camera issues with YOLO models. Try using `picamera2` instead of OpenCV's VideoCapture, which comes pre-installed with Raspberry Pi OS and works better with the libcamera system:

```python
import cv2
from picamera2 import Picamera2
from ultralytics import YOLO

# Initialize the Picamera2
picam2 = Picamera2()
picam2.preview_configuration.main.size = (1280, 720)
picam2.preview_configuration.main.format = "RGB888"
picam2.preview_configuration.align()
picam2.configure("preview")
picam2.start()

# Load the YOLO model
model = YOLO("best.pt")

while True:
    # Capture frame-by-frame
    frame = picam2.capture_array()

    # Run YOLO inference on the frame
    results = model(frame, conf=0.5)

    # Visualize the results on the frame
    annotated_frame = results[0].plot()

    # Display the resulting frame
    cv2.imshow("Blueberry Detection", annotated_frame)

    # Break the loop if 'q' is pressed
    if cv2.waitKey(1) == ord("q"):
        break

cv2.destroyAllWindows()
```

This approach is specifically designed for Raspberry Pi cameras and should work reliably with your setup.

**Comment by Shamsu992** [comment]
Wow, I will try this out tomorrow Glenn and post my progress in this chat, hopefully GitHub does not close the chat, thank you and I will give it a try first thing tomorrow

**Comment by glenn-jocher** [comment]
@Shamsu992 sounds great! The `picamera2` approach should resolve your camera issues since it's specifically designed for Raspberry Pi's libcamera system. Feel free to report back here with your results, and don't worry - the issue will stay open for your update.
